[Features][CustomSched][BPFLib]

abrehman94 opened the pull request:
* Custom Scheduling
    * Capturing cgroup stats on per invoke level
    * BPFLibrary to create a pinned BPFMap of function characteristics
      * used to share data with scx scheduler

aFuerst commented on src/Ilúvatar/Makefile:
```diff
@@ -1,6 +1,6 @@
 NPROCS:=$(shell grep -c ^processor /proc/cpuinfo)
 CROSS_FLAGS=CROSS_CONTAINER_UID=0 CROSS_CONTAINER_GID=0 CROSS_ROOTLESS_CONTAINER_ENGINE=false CROSS_CONTAINER_USER_NAMESPACE=none CROSS_CONTAINER_OPTS=--privileged
-RUST_FLAGS=-Dwarnings
+#RUST_FLAGS=-Dwarnings
```
We don't want any warnings in the mainline, so once this is in a ready state, re-enable this please.

abrehman94 replied:
Re-enabled.

aFuerst commented on src/Ilúvatar/Makefile:
```diff
@@ -55,4 +58,4 @@ format:
 	@cargo fmt --all -- --check
 clippy:
-	@RUSTFLAGS=$(RUST_FLAGS) $(RUST_C) clippy --workspace --examples --benches --no-deps -- -Dclippy::suspicious -Dclippy::correctness -Dclippy::perf -Aclippy::single_match -Aclippy::new_without_default -Aclippy::too_many_arguments -Aclippy::type-complexity -Dclippy::from_over_into -Aclippy::redundant-field-names -Dwarnings
+	@RUSTFLAGS=$(RUST_FLAGS) $(RUST_C) clippy --fix --allow-no-vcs --workspace --examples --benches --no-deps -- -Dclippy::suspicious -Dclippy::correctness -Dclippy::perf -Aclippy::single_match -Aclippy::new_without_default -Aclippy::too_many_arguments -Aclippy::type-complexity -Dclippy::from_over_into -Aclippy::redundant-field-names -Dwarnings
```
This command is used to _check_ if there are clippy violations, especially in CI, not to correct them. Adding `--fix` will have the issues "fixed" in the CI build, but those changes aren't actually committed to this branch. The issues would persist and silently creep in code quality problems. Undo this

abrehman94 replied:
fixed

aFuerst commented on src/Ilúvatar/Makefile:
```diff
@@ -37,6 +37,9 @@ tiny:
 	@RUSTFLAGS=$(RUST_FLAGS) $(RUST_C) build $(DEBUG_FLAGS) $(CARGO_ARGS)
+fix:
+	@echo "Fixing lint errors"
+	@RUSTFLAGS=$(RUST_FLAGS) $(RUST_C) fix $(DEBUG_FLAGS) $(CARGO_ARGS)
```
Combine this with the clippy fix command you changed below? Would be nice to have in the makefile to simplify fixing them locally

abrehman94 replied:
done

aFuerst commented on src/Ilúvatar/Cross.toml:
```diff
@@ -5,5 +5,9 @@ passthrough = ["ARCH=amd64", "GO_VERSION=1.22.0", "CNI_VERSION=v1.1.1", "GOPATH=
     "dpkg --add-architecture $CROSS_DEB_ARCH", 
-    "apt-get update && apt-get --assume-yes install protobuf-compiler iproute2 wget curl runc bridge-utils iptables net-tools sysstat"
+    "apt-get update && apt-get --assume-yes install protobuf-compiler iproute2 wget curl runc bridge-utils iptables net-tools sysstat libelf-dev lsb-release wget software-properties-common gnupg",
+    "wget https://apt.llvm.org/llvm.sh && chmod +x llvm.sh &&  ./llvm.sh  all &&  ln -s /usr/bin/ld.lld-18 /usr/bin/ld.lld &&  ln -s /usr/bin/clang-18 /usr/bin/clang &&  ln -s /usr/bin/clang++-18 /usr/bin/clang++ && rm llvm.sh"
```
Can you add these to the setup docs as new dependencies? I assume they're required for the bpf stuff

abrehman94 replied:
These are required at the build time. From what I understand, cross builds in a container environment so these dependencies need to be installed there. It does not need to be on the host. Setup docs are for the host right?

aFuerst commented on src/Ilúvatar/fs_policy_tsksz/src/bpf.rs:
```diff
@@ -0,0 +1,219 @@
+        let rb_map = &mut skel.maps.queued_stats;
+        let mut builder = libbpf_rs::RingBufferBuilder::new();
+        builder.add(rb_map, callback_stats).unwrap();
+        let queued_stats = builder.build().unwrap();
```
We don't want `upwrap` and `panic!` because they can cause hard-to-debug issues and the error is only sent to stderr and easily lost

abrehman94 replied:
fs_policy_tsksz is launched as a separate process and the stderr is being logged to a separate log file. In case of a panic log file can be interpreted offline.

aFuerst commented on src/Ilúvatar/fs_policy_tsksz/src/main.rs:
```diff
@@ -0,0 +1,213 @@
+            loop {
+                match crecvs.rx_chr.try_recv() {
+                    Ok(chr) => {
+                        // Do something interesting with your result
```
Remove all the any `println` , testing comments, etc.

abrehman94 replied:
Removed testing comments and unnecessary println. stdout of fs_policy is being logged to a separate file and some println are important.

aFuerst commented on src/Ilúvatar/fs_policy_tsksz/src/main.rs:
```diff
@@ -0,0 +1,213 @@
+                }
+            }
+            for (k, v) in &self.characteristics {
+                println!("{}: {:?}", k, v);
```
Idk how this will run, but can it use the same logging framework we have? For consistency

abrehman94 replied:
It's running as a separate process. I am logging it's stdout and stderr to separate files. We can use same tracing framework but it would just change the logging format in those files. 

Not sure if you meant logging to same file or just logging using this framework for consistent format.

aFuerst commented on src/Ilúvatar/iluvatar_library/src/cgroup_interaction.rs:
Is there am existing crate to make cgroup interaction easier on us and avoid duplication? I trust it's doing what you want, don't know what's out there.

abrehman94 replied:
Closest I found was this https://crates.io/crates/cgroups-fs. It does not support cgroup v2.

aFuerst commented on src/Ilúvatar/iluvatar_library/src/cgroup_interaction.rs:
```diff
@@ -0,0 +1,656 @@
+                procs: vec![],
+                cpustats: [
+                    ("nr_periods".to_string(), 0),
+                    ("nr_throttled".to_string(), 0),
```
These strings should be constants

aFuerst commented on src/Ilúvatar/iluvatar_library/src/cgroup_interaction.rs:
```diff
@@ -0,0 +1,656 @@
+        let cgroup_id = "89e979a2b0e9fd3".to_string();
+        let res = read_cgroup(cgroup_id);
+        println!("{:?}", res);
+        assert!(false);
```
Get these tests working, and remove print statements in them please.

aFuerst commented on src/Ilúvatar/iluvatar_library/src/utils/mod.rs:
```diff
@@ -17,10 +17,38 @@ use async_process::Command as AsyncCommand;
+        .arg("-c")
+        .arg(cmd)
+        .output()
+        .expect("failed to execute process");
```
There are wrappers for executing commands, `iluvatar_library::utils::execute_cmd` and others

abrehman94 replied:
It was experimental code. I have removed it.

aFuerst commented on src/Ilúvatar/iluvatar_worker_library/src/services/containers/container_pool.rs:
```diff
@@ -111,6 +121,8 @@ impl ContainerPool {
                 Some(c) => {
                     debug!(tid=%tid, container_id=%c.container_id(), name=%self.pool_name, pool_type=?PoolType::Idle, "Removing random container from pool");
                     self.add_container(c.clone(), &self.running_pool, tid, PoolType::Running);
+                    // pool is the best place to maintain a tid to cgroup id mapping
```
Is this the best place? Presumably the cgroup on a container doesn't change and could just be stored in the container struct. This would just make maintaining the pool and cgroups more convoluted

abrehman94 replied:
There was no mapping for tid to container. It creates this map in the container pool. 

I don't think during a transaction tid a cgroup-id can change. Correct me if I am wrong. 

Cgroup-id is part of the container.

aFuerst commented on src/Ilúvatar/iluvatar_worker_library/src/services/invocation/cpu_q_invoke.rs:
```diff
@@ -333,6 +333,8 @@ impl CpuQueueingInvoker {
         };
         self.running.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+        // since the acquire has been called - tid-cgroup id should be there
+        self.cmap.start_invoke(&reg.fqdn, tid);
```
Put these in `invoke_on_container_2`, one combined place that already has the cmap in it, much less code threading you have to do

abrehman94 replied:
That is a very nice suggestion! moved it to invoke_on_container_2.

aFuerst commented on src/Ilúvatar/iluvatar_worker_library/src/worker_api/mod.rs:
```diff
@@ -25,8 +35,112 @@ pub mod rpc;
+    pub tx_pids: IpcSender<PidsPacket>,
+}
+
+fn launch_scheduler(worker_config: &WorkerConfig) {
```
All this should be in a separate file. It is much too specific to be in the startup code file

abrehman94 replied:
Moved to a separate file.

aFuerst commented:
How much did you change with that push? You re-wrote the history so I can't really tell. And please don't do that again...

Is there more that needs done to this?

abrehman94 commented:
I was trying to keep the history clean. Specific changes made by the force push can be viewed using [compare button](https://github.com/COS-IN/iluvatar-faas/compare/939b86bdc4f51991c29f3254dd1b536eac54cc70..03aa34f437adbc653f4226dd39cabd1999394c5c).

Last commit does not fix everything. I have removed unnecessary code and fixed warnings only. 

Right now I am on vacation, I will get to it in a week.

abrehman94 commented:
@aFuerst Please have a look. I think I have addressed all the review comments. Sorry about force pushing the branch. It's just too much code, if I would have done separate commits, it would pollute the git history.
